### PR TITLE
E2E: Add regression test for custom mqtt topics

### PR DESCRIPTION
**Pull request by and-rewsmith**
This PR is adding an E2E test for custom mqtt topics using the existing ```genericMqttTester``` image we run in longhaul. 

This test will deploy edgehub with the broker enabled and two genericMqttTester modules. One genericMqttTester module will be in initiate mode and the other will be in relay mode. The initiate mode module will publish on an mqtt topic ```initiate/1``` and the relaying module will subscribe to this topic. When the relaying module receives a message, it will publish it back on ```relay/1```. The initiating module will be subscribed on this ```relay/1``` topic and will report to the TRC when it receives the message.

In addition to adding the test it:
1. Refactors shared logic into util abstractions for cleaner code. Specifically adding the broker/TRC to the deployment.
2. Introduces a configuration for defining the restartPolicy for test module deployed in the test. This is done so that the genericMqttTester module will no start back up after it sends its defined amount of messages.
3. Improves the genericMqttTester module to use the module name when reporting to the trc rather than some hardcoded string "genericMqttTester". This allows the dotnet test code we right here to be more understandable.
3. Updates the documentation to include new parameters required for this test

### CODE REVIEW

File: test/Microsoft.Azure.Devices.Edge.Test.Common/config/EdgeConfigBuilder.cs [hunk 1]
```diff
@@ -45,9 +45,9 @@ public IModuleConfigBuilder AddEdgeHub(string image = null, bool optimizeForPerf
             return builder;
         }
 
-        public IModuleConfigBuilder AddModule(string name, string image)
+        public IModuleConfigBuilder AddModule(string name, string image, bool shouldRestart = true)
```
Comment (and-rewsmith) [inline test/Microsoft.Azure.Devices.Edge.Test.Common/config/EdgeConfigBuilder.cs:48]: I am defaulting the restart policy to "always" as it is the intention for the ~20 occurrences where this function is called.

**Comment by and-rewsmith** [comment]
Passing E2E:
https://msazure.visualstudio.com/One/_build/results?buildId=41593376&view=ms.vss-test-web.build-test-results-tab

The GetModuleLogs passed on a rerun, but this test was not affected by this PR's changes.

File: test/Microsoft.Azure.Devices.Edge.Test/GenericMqtt.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Test.Common;
+    using Microsoft.Azure.Devices.Edge.Test.Common.Config;
+    using Microsoft.Azure.Devices.Edge.Test.Helpers;
+    using Microsoft.Azure.Devices.Edge.Util.Test.Common.NUnit;
+    using NUnit.Framework;
+
+    [EndToEnd]
```
Comment (nlcamp) [inline test/Microsoft.Azure.Devices.Edge.Test/GenericMqtt.cs:13]: Adding a comment block here with some of the text from the description section of this PR might be useful here to document what this class does.

**Review by nlcamp — APPROVED** [review]

**Comment by and-rewsmith** [comment]
There is an existing problem in nested e2e tests where broker specific tests run in a non-broker topology and fail because this configuration doesn't work. This affects the test I just added and other tests too (BridgeConfigUpdate). I think the right approach is to merge this PR and fix this unrelated bug in a separate PR. Then everything will pass.

Specifically, the fix I have in mind will add a field to Context.Current that indicates the upstream protocol. That way we can not run broker-specific tests in a non-broker topology.

**Pull request merged**